File: golang-devildyw/middle/L24两两交换链表中的节点.go
package main

func main() {

}

func swapPairs(head *ListNode) *ListNode {

	var (
		dummyNode = &ListNode{
			Next: head,
		}
		cur = dummyNode
	)

	// 画图模拟即可
	for cur.Next != nil && cur.Next.Next != nil {
		//需要提前保存一些节点 交换的时候 因为交换期间可能会改变链路
		tempNext := cur.Next
		tempNext1 := cur.Next.Next.Next
		cur.Next = cur.Next.Next
		cur.Next.Next = tempNext
		tempNext.Next = tempNext1
		cur = tempNext
	}

	return dummyNode.Next
}

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L24两两交换链表中的节点.java
package top.dyw.leetcode.Hot100;

public class L24两两交换链表中的节点 {
    public ListNode swapPairs(ListNode head) {
        ListNode dummyNode = new ListNode(-1, head);
        ListNode pre = dummyNode;
        ListNode cur = dummyNode;

        while(true) {
            int n = 2;
            while(n>0 && cur != null) {
                cur = cur.next;
                n--;
            }

            if (cur==null) {
                break;
            }

            //开始反转
            ListNode next = cur.next;
            cur.next = null;

            ListNode tail = pre.next;
            pre.next = reverseList(tail);

            tail.next = next;
            cur = tail;
            pre = tail;
        }

        return dummyNode.next;
    }

    public ListNode reverseList(ListNode tail) {
        ListNode head = null;
        while(tail != null) {
            ListNode next = tail.next;
            tail.next = head;
            head = tail;
            tail = next;
        }
        return head;
    }
}
